### PR TITLE
Metrics per endpoint

### LINKED ISSUE #119: Per-endpoint ring metrics
Seems like it would be useful to be able to record metrics for each endpoint in an app. In theory since it just wraps a handler, you could wrap each endpoint individually, but then you'd register the same metric twice. Would be nice if the "instrument-ring" function accepted a prefix that could be used to ensure each metric is uniquely named when wrapping multiple endpoints.

Does that make sense? I can come up with a PR if this sounds ok.

### PR DESCRIPTION
**Pull request by cddr**
Closes #119 

Update:

Added tests that actually check the expected metrics have been registered and extracted the bit that adds them into it's own function since cider was complaining that the method was too big to debug. I could put that back now that I'm finished debugging but it seems like a reasonable refactor in any case.

### CODE REVIEW

**Comment by cddr** [comment]
Actually I think this needs a bit of a rethink. I tried to integrate it with our own app and it was very inconvenient to have to wrap each individual endpoint. I'm going to make a change that will allow you to just wrap a single handler but still track metrics per endpoint.

**Comment by michaelklishin** [comment]
@cddr thank you! Having this kind of feedback before a PR is merged is incredibly valuable.

File: metrics-clojure-ring/src/metrics/ring/instrument.clj [hunk 1]
```diff
@@ -2,62 +2,107 @@
...
   (:import [com.codahale.metrics MetricRegistry]))
 
+(def default-options
+  {:prefix []})
 
 (defn- mark-in! [metric-map k]
   (when-let [metric (metric-map k (metric-map :other))]
     (mark! metric)))
 
+(defn- ring-metrics [reg {:keys [prefix] :as options}]
```
Comment (michaelklishin) [inline metrics-clojure-ring/src/metrics/ring/instrument.clj]: Sorry to nitpick but can you please move the argument vector to the next line? We don't do this consistently in this project
but every function that has a doc string or has multiple heads (overrides) does that.

**Comment by cddr** [comment]
Hi Michael,

Thanks for looking at this so quickly. Apologies for re-opening it before finishing the docs. That's me done now. I put the arglist on it's own line as you wished :-)

I am a little concerned at the risk this introduces for unbounded memory growth (an attacker could hit the app with many unique uris and we'd create the metrics map for each distinct one). However I can't think of an easy way to limit this only to URIs that would actually be served by the app. This isn't a problem for us but could be if you have a public facing website/api.

I suppose this is one of the benefits of the router being expressed as data rather than code (e.g. bidi instead of compojure).

**Comment by michaelklishin** [comment]
Thank you!

**Pull request merged**